File: Simulator/app/src/main/java/com/example/simulator/ui/MainActivity.java
package com.example.simulator.ui;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.simulator.R;
import com.example.simulator.data.MatchesApi;
import com.example.simulator.databinding.ActivityMainBinding;
import com.example.simulator.domain.Match;
import com.example.simulator.ui.adapter.MatchAdapterJAVA;
import com.google.android.material.snackbar.Snackbar;

import java.util.Collections;
import java.util.List;
import java.util.Random;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private MatchesApi matchesApi;
    private MatchAdapterJAVA matchesAdapter = new MatchAdapterJAVA(Collections.EMPTY_LIST);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setupHttpClient();
        setupMatchesList();
        setupMatchesRefresh();
        setupFloatingActionButton();
    }


    private void setupHttpClient() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://rianlimeira.github.io/matches-simulator-api/")
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        matchesApi = retrofit.create(MatchesApi.class);
    }

    private void setupMatchesList() {
        /* TODO LISTAR AS PARTIDAS, CONSUMINDO API */
        binding.rvMatches.setHasFixedSize(true);
        binding.rvMatches.setLayoutManager(new LinearLayoutManager(this));

        binding.rvMatches.setAdapter(matchesAdapter);
        findMatchesFromApi();

    }

    private void setupMatchesRefresh() {
        /* TODO: ATUALIZAR AS PARTIDAS NA A&Ccedil;&Atilde;O DE SWIPE */
        binding.srlMatches.setOnRefreshListener(this::findMatchesFromApi);
    }

    private void setupFloatingActionButton() {
        /* TODO: CRIAR EVENTO DE CLICK E SIMULÇÃO */
        binding.fabSimulator.setOnClickListener(view -> {
            view.animate().rotationBy(360).setDuration(500).setListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationCancel(Animator animation) {
                    /* TODO Implementar o algoritimo de simulação de partidas */
                    Random random = new Random();
                    for (int i = 0; i< matchesAdapter.getItemCount(); i++){
                        Match match = matchesAdapter.getMatches().get(i);
                        match.getHomeTeam().setScore(random.nextInt(match.getHomeTeam().getStars() +1));
                        match.getAwayTeam().setScore(random.nextInt(match.getAwayTeam().getStars() +1));
                        matchesAdapter.notifyItemChanged(i);
                    }
                }
            });
            //Teste para crashlystics, forçando parada
            //throw new RuntimeException("Teste Crash");
        });
    }

    private void showErrorManager() {
        Snackbar.make(binding.fabSimulator, R.string.error_api, Snackbar.LENGTH_LONG).show();
    }

    private void findMatchesFromApi() {
        binding.srlMatches.setRefreshing(true);
        matchesApi.getMatches().enqueue(new Callback<List<Match>>() {
            @Override
            public void onResponse(@NonNull Call<List<Match>> call, @NonNull Response<List<Match>> response) {
                if (response.isSuccessful()){
                    List<Match> matches = response.body();
                    matchesAdapter = new MatchAdapterJAVA(matches);
                    binding.rvMatches.setAdapter(matchesAdapter);
                }
                else {
                    showErrorManager();
                }
                binding.srlMatches.setRefreshing(false);
            }

            @Override
            public void onFailure(@NonNull Call<List<Match>> call, @NonNull Throwable t) {
                showErrorManager();
                binding.srlMatches.setRefreshing(false);
            }
        });
    }
}

File: Simulator/app/src/main/java/com/example/simulator/ui/MainActivity.kt
package com.example.simulator.ui

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.example.simulator.R
import com.example.simulator.data.MatchesApi
import com.example.simulator.databinding.ActivityMainBinding
import com.example.simulator.domain.Match
import com.google.android.material.snackbar.Snackbar
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class MainActivity : AppCompatActivity() {
    private lateinit var matchesApi: MatchesApi
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        /*setContentView(R.layout.activity_main)
        val textHello = findViewById<TextView>(R.id.hello)*/

        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupHTTPCliente()
        setupMatchesList()
        setupMatchesRefresh()
        setupFloatingActionButton()

    }

    private fun setupHTTPCliente() {
        val retrofit = Retrofit.Builder()
            .baseUrl("https://rianlimeira.github.io/matches-simulator-api/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        matchesApi = retrofit.create(MatchesApi::class.java)
    }

    private fun setupMatchesList() {
        //TODO LISTAR AS PARTIDAS, CONSUMINDO API

        //TODO LISTAR AS PARTIDAS, CONSUMINDO API
        matchesApi.matches.enqueue(object : Callback<List<Match?>> {
            override fun onResponse(call: Call<List<Match?>>, response: Response<List<Match?>>) {
                if (response.isSuccessful) {
                    val matches = response.body()!!
                    Log.i("SIMULATOR", "Deu tudo certo! Partidas = " + matches.size)
                } else {
                    showErrorManager()
                }
            }

            override fun onFailure(call: Call<List<Match?>>, t: Throwable) {
                showErrorManager()
            }
        })
    }

    private fun setupFloatingActionButton() {
        //TODO: CRIAR EVENTO DE CLICK E SIMULÇÃO
    }


    private fun setupMatchesRefresh(){
        //TODO: ATUALIZAR AS PARTIDAS NA AÇÃO DE SWIPE
    }

    private fun showErrorManager() {
        Snackbar.make(binding.fabSimulator, R.string.error_api, Snackbar.LENGTH_LONG).show()
    }
}

